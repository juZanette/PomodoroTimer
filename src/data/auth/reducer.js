const initialState = {
  signed: false,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SIGN_IN':
      return {
        signed: true,
      }
    default:
      return state
  }
}
