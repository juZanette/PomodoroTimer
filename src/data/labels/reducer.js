const initialState = {
  dialogOpened: false,
  labelToEdit: null,
  formValue: {
    name: null,
    color: null,
  },
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
