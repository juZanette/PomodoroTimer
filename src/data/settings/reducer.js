const defaultState = {
  workDuration: 25,
  shortBreakDuration: 5,
  longBreakDuration: 20,
  rounds: 4,
  showTimerInTitle: true,
  showNotifications: true,
}

const initialState = {
  workDuration: null,
  shortBreakDuration: null,
  longBreakDuration: null,
  rounds: null,
  showTimerInTitle: null,
  showNotifications: null,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_WORK_DURATION':
      return {
        ...state,
        workDuration: action.duration,
      }
    case 'SET_SHORT_BREAK_DURATION':
      return {
        ...state,
        shortBreakDuration: action.duration,
      }
    case 'SET_LONG_BREAK_DURATION':
      return {
        ...state,
        longBreakDuration: action.duration,
      }
    case 'SET_ROUNDS':
      return {
        ...state,
        rounds: action.rounds,
      }
    case 'SET_SHOW_TIMER_IN_TITLE':
      return {
        ...state,
        showTimerInTitle: action.showTimerInTitle,
      }
    case 'SET_SHOW_NOTIFICATIONS':
      return {
        ...state,
        showNotifications: action.showNotifications,
      }
    case 'SET_SETTINGS':
      return {
        ...defaultState,
        ...action.settings,
      }

    default:
      return state
  }
}
