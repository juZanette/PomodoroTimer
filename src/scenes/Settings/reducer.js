const initialState = {
  rounds: 4,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_ROUNDS':
      return {
        rounds: action.roundsNumber,
      }
    default:
      return state
  }
}
