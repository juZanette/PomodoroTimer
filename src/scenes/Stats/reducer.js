const initialState = {
  totalSessions: 0,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'INCREMENT_TOTAL_SESSIONS':
      return {
        totalSessions: ++state.totalSessions,
      }
    default:
      return state
  }
}
