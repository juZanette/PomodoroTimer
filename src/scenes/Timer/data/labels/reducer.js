const initialState = {
  dialogOpened: false,
  formDialogOpened: false,
  deleteAlert: {
    opened: false,
    labelToDelete: null,
  },
  labelToEdit: null,
  formValue: null,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_FORM_DIALOG':
      return {
        ...state,
        formDialogOpened: action.formDialogOpened,
      }
    case 'SET_DELETE_ALERT':
      const { opened, labelToDelete } = action.deleteAlert

      return {
        ...state,
        deleteAlert: {
          opened,
          labelToDelete,
        },
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
