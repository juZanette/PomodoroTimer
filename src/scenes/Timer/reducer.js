const initialState = {
  timerActive: false,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_TIMER_ACTIVE':
      return {
        timerActive: action.timerActive,
      }
    default:
      return state
  }
}
