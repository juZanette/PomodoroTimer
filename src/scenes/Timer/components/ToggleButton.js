import React from 'react'
import Box from '@material-ui/core/Box'
import IconButton from '@material-ui/core/IconButton'
import styled from 'styled-components'
import PlayArrowIcon from '@material-ui/icons/PlayArrow'
import PauseIcon from '@material-ui/icons/Pause'
import { useDispatch, useSelector } from 'react-redux'
import {
  setNextTimer,
  setStatus,
  setTimeLeft,
  setProgress,
  pauseTimer,
  saveInterval,
} from '../data/timer/actions'
import { STATUSES, TYPES, setTitle } from '../data/timer/reducer'
import chime from '../assets/chime.mp3'
import work from '../assets/work.png'
import alarm from '../assets/alarm.png'
import coffee from '../assets/coffee.png'
import { startAddSession } from '../../../data/sessions/actions'

const ActionIcon = styled(IconButton)`
  border: 1px solid #bababa;
`

export const ToggleButton = () => {
  const { status, timeLeft, type } = useSelector((state) => state.timer)
  const settings = useSelector((state) => state.settings)
  const label = useSelector((state) => state.labels.labelSelected)

  const dispatch = useDispatch()

  const audio = new Audio(chime)

  const calculateTimeLeft = (endTime) => {
    const difference = +endTime - +new Date()
    let timeLeft = { minutes: 0, seconds: 0 }

    if (difference > 0) {
      timeLeft = {
        minutes: Math.floor((difference / 1000 / 60) % 60),
        seconds: Math.floor((difference / 1000) % 60),
      }
    }

    return timeLeft
  }

  const calculateProgress = (duration, timeLeft) => {
    const secondsTotal = duration * 60
    const secondsLeft = timeLeft.minutes * 60 + timeLeft.seconds

    return 100 / (secondsTotal / secondsLeft)
  }

  const startTimer = () => {
    if (status === STATUSES.running) return

    if (settings.showNotifications && 'Notification' in window) {
      Notification.requestPermission()
    }

    dispatch(setStatus(STATUSES.running))

    const endTime = new Date(
      new Date().getTime() + timeLeft.minutes * 60000 + timeLeft.seconds * 1000
    )

    const duration = timeLeft.minutes

    const interval = setInterval(() => {
      const calculatedTimeLeft = calculateTimeLeft(endTime)
      const calculatedProgress = calculateProgress(duration, calculatedTimeLeft)

      dispatch(setTimeLeft(calculatedTimeLeft))
      dispatch(setProgress(calculatedProgress))

      if (settings.showTimerInTitle) {
        setTitle(type, calculatedTimeLeft)
      }

      if (!calculatedProgress) {
        dispatch(
          startAddSession({
            label: label ? label.id : null,
            duration: { minutes: settings.workDuration, seconds: 0 },
          })
        )

        setTimeout(async () => {
          dispatch(setNextTimer(settings))
          audio.play()

          if (
            settings.showNotifications &&
            'Notification' in window &&
            Notification.permission === 'granted'
          ) {
            const msg =
              type === TYPES.work ? 'Take a break ‚òïÔ∏è' : 'Start working üë®‚Äçüíª'

            const icon = type === TYPES.work ? coffee : work

            const registration = await navigator.serviceWorker.ready

            registration.showNotification(msg, {
              vibrate: [100, 50, 100],
              badge: alarm,
              icon,
            })
          }
        }, 1000)

        clearInterval(interval)
      }
    }, 200)

    dispatch(saveInterval(interval))
  }

  return (
    <Box display="flex" justifyContent="center" mt={2}>
      {status !== STATUSES.running && (
        <ActionIcon
          disabled={!timeLeft}
          aria-label="start timer"
          onClick={startTimer}
        >
          <PlayArrowIcon />
        </ActionIcon>
      )}

      {status === STATUSES.running && (
        <ActionIcon
          aria-label="pause timer"
          onClick={() => {
            dispatch(pauseTimer())
          }}
        >
          <PauseIcon />
        </ActionIcon>
      )}
    </Box>
  )
}
