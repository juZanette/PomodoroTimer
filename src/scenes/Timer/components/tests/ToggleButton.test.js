import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { createMount } from '@material-ui/core/test-utils'
import IconButton from '@material-ui/core/IconButton'
import { ToggleButton } from '../ToggleButton'
import { STATUSES, TYPES } from '../../data/timer/reducer'

describe('<ToggleButton />', () => {
  const mockStore = configureMockStore([thunk])

  let store
  let mount
  let wrapper

  beforeEach(() => {
    mount = createMount()

    wrapper = (status = STATUSES.onHold) => {
      const storeData = {
        timer: {
          status,
          type: TYPES.work,
          timeLeft: { minutes: 25, seconds: 0 },
        },
        settings: {},
        labels: { labelSelected: null },
      }

      store = mockStore(storeData)
      store.dispatch = jest.fn()

      return mount(
        <Provider store={store}>
          <ToggleButton />
        </Provider>
      )
    }
  })

  test('should render <ToggleButton /> correctly', () => {
    expect(wrapper()).toMatchSnapshot()
  })

  test('should render Pause button if the timer is running', () => {
    const wrapperRendered = wrapper(STATUSES.running)
    expect(wrapperRendered.find(IconButton).prop('aria-label')).toBe(
      'pause timer'
    )
  })

  test('should render Start button if the timer not running', () => {
    const wrapperRendered = wrapper()
    expect(wrapperRendered.find(IconButton).prop('aria-label')).toBe(
      'start timer'
    )
  })
})
