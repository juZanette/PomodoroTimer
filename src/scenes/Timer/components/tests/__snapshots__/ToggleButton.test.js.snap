// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButton /> should render <ToggleButton /> correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ToggleButton>
    <Styled(MuiBox)
      display="flex"
      justifyContent="center"
      mt={2}
    >
      <div
        className="MuiBox-root MuiBox-root-2"
      >
        <Styled(WithStyles(ForwardRef(IconButton)))
          aria-label="start timer"
          disabled={false}
          onClick={[Function]}
        >
          <WithStyles(ForwardRef(IconButton))
            aria-label="start timer"
            className="sc-AxjAm jVitc"
            disabled={false}
            onClick={[Function]}
          >
            <ForwardRef(IconButton)
              aria-label="start timer"
              className="sc-AxjAm jVitc"
              classes={
                Object {
                  "colorInherit": "MuiIconButton-colorInherit",
                  "colorPrimary": "MuiIconButton-colorPrimary",
                  "colorSecondary": "MuiIconButton-colorSecondary",
                  "disabled": "Mui-disabled",
                  "edgeEnd": "MuiIconButton-edgeEnd",
                  "edgeStart": "MuiIconButton-edgeStart",
                  "label": "MuiIconButton-label",
                  "root": "MuiIconButton-root",
                  "sizeSmall": "MuiIconButton-sizeSmall",
                }
              }
              disabled={false}
              onClick={[Function]}
            >
              <WithStyles(ForwardRef(ButtonBase))
                aria-label="start timer"
                centerRipple={true}
                className="MuiIconButton-root sc-AxjAm jVitc"
                disabled={false}
                focusRipple={true}
                onClick={[Function]}
              >
                <ForwardRef(ButtonBase)
                  aria-label="start timer"
                  centerRipple={true}
                  className="MuiIconButton-root sc-AxjAm jVitc"
                  classes={
                    Object {
                      "disabled": "Mui-disabled",
                      "focusVisible": "Mui-focusVisible",
                      "root": "MuiButtonBase-root",
                    }
                  }
                  disabled={false}
                  focusRipple={true}
                  onClick={[Function]}
                >
                  <button
                    aria-label="start timer"
                    className="MuiButtonBase-root MuiIconButton-root sc-AxjAm jVitc"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="button"
                  >
                    <span
                      className="MuiIconButton-label"
                    >
                      <ForwardRef>
                        <WithStyles(ForwardRef(SvgIcon))>
                          <ForwardRef(SvgIcon)
                            classes={
                              Object {
                                "colorAction": "MuiSvgIcon-colorAction",
                                "colorDisabled": "MuiSvgIcon-colorDisabled",
                                "colorError": "MuiSvgIcon-colorError",
                                "colorPrimary": "MuiSvgIcon-colorPrimary",
                                "colorSecondary": "MuiSvgIcon-colorSecondary",
                                "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                "root": "MuiSvgIcon-root",
                              }
                            }
                          >
                            <svg
                              aria-hidden={true}
                              className="MuiSvgIcon-root"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M8 5v14l11-7z"
                              />
                            </svg>
                          </ForwardRef(SvgIcon)>
                        </WithStyles(ForwardRef(SvgIcon))>
                      </ForwardRef>
                    </span>
                    <WithStyles(memo)
                      center={true}
                    >
                      <ForwardRef(TouchRipple)
                        center={true}
                        classes={
                          Object {
                            "child": "MuiTouchRipple-child",
                            "childLeaving": "MuiTouchRipple-childLeaving",
                            "childPulsate": "MuiTouchRipple-childPulsate",
                            "ripple": "MuiTouchRipple-ripple",
                            "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                            "rippleVisible": "MuiTouchRipple-rippleVisible",
                            "root": "MuiTouchRipple-root",
                          }
                        }
                      >
                        <span
                          className="MuiTouchRipple-root"
                        >
                          <TransitionGroup
                            childFactory={[Function]}
                            component={null}
                            exit={true}
                          />
                        </span>
                      </ForwardRef(TouchRipple)>
                    </WithStyles(memo)>
                  </button>
                </ForwardRef(ButtonBase)>
              </WithStyles(ForwardRef(ButtonBase))>
            </ForwardRef(IconButton)>
          </WithStyles(ForwardRef(IconButton))>
        </Styled(WithStyles(ForwardRef(IconButton)))>
      </div>
    </Styled(MuiBox)>
  </ToggleButton>
</Provider>
`;
